### PR TITLE
feat: changing to use EKS for management cluster

### LINKED ISSUE #382: [E2E] Fix Nightly tests
The nightly (i.e. long) e2e tests have been failing for some time now. For example:

https://github.com/rancher-sandbox/rancher-turtles/actions/runs/7880148289

We need to fix them and also remove the usage of ngrok. Lets make the following changes:

- Create an EKS based Rancher Manager cluster
- Use nginx ingress and a AWS LB for the host
- Update the Rancher install to use the AWS LB

We could potentially use eksctl to make the provisioning easier.

This would also require #368 to ensure that we clean up.

### PR DESCRIPTION
**Pull request by richardcase**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #382 

**Special notes for your reviewer**:

**Checklist**:

- [x] squashed commits into logical changes
- [ ] includes documentation
- [ ] adds unit tests
- [x] adds or updates e2e tests

### CODE REVIEW

**Comment by Danil-Grigorev** [comment]
Can you please rebase @richardcase? Also, does this include a fix for failing nightlies with the v1alpha3 CAPI version issue? Or there is a separate PR for that/needs to be yet done?

**Comment by richardcase** [comment]
> Can you please rebase @richardcase? Also, does this include a fix for failing nightlies with the v1alpha3 CAPI version issue? Or there is a separate PR for that/needs to be yet done?

Yes i will rebase @Danil-Grigorev , just trying to fix the "quick" tests after the refactor. It contains the fix for the nightly failures as well

File: Makefile [hunk 1]
```diff
@@ -519,12 +520,17 @@ test-e2e: $(GINKGO) $(HELM) $(CLUSTERCTL) kubectl e2e-image ## Run the end-to-en
...
+		-e2e.use-eks=$(USE_EKS)
 
 .PHONY: e2e-image
 e2e-image: ## Build the image for e2e tests
-	TAG=v0.0.1 $(MAKE) docker-build
-	RELEASE_TAG=v0.0.1 CONTROLLER_IMG=$(MANIFEST_IMG) CONTROLLER_IMAGE_VERSION=v0.0.1 $(MAKE) build-chart
+	TAG=v0.0.1 CONTROLLER_IMAGE_NAME=rancher-turtles-e2e $(MAKE) docker-build
+	RELEASE_TAG=v0.0.1 CONTROLLER_IMG=$(REGISTRY)/$(ORG)/rancher-turtles-e2e-$(ARCH) CONTROLLER_IMAGE_VERSION=v0.0.1 $(MAKE) build-chart
+
+.PHONY: e2e-image-pus
```
[the remark this answers is not in the document]
Comment (richardcase) [reply Makefile:531]: changed.

File: charts/rancher-turtles/templates/pre-install-job.yaml [hunk 2]
```diff
@@ -67,7 +67,7 @@ spec:
       serviceAccountName: pre-install-job
       containers:
         - name: rancher-mutatingwebhook-cleanup
-          image: {{ index .Values "rancherTurtles" "features" "rancher-webhook" "kubectlImage" }}:{{ template "strippedKubeVersion" . }}
+          image: {{ index .Values "rancherTurtles" "features" "rancher-webhook" "kubectlImage" }}:{{ template "kubectlImageTag" . }}
```
[the remark this answers is not in the document]
Comment (richardcase) [reply charts/rancher-turtles/templates/pre-install-job.yaml:70]: Removed

File: charts/rancher-turtles/values.yaml [hunk 3]
```diff
@@ -1,7 +1,7 @@
 rancherTurtles:
   image: controller
   imageVersion: v0.0.0
-  imagePullPolicy: Never
+  imagePullPolicy: IfNotPresent
```
[the remark this answers is not in the document]
Comment (richardcase) [reply charts/rancher-turtles/values.yaml:4]: Done :)

File: test/testenv/rancher.go [hunk 4]
```diff
@@ -196,6 +195,9 @@ func DeployRancher(ctx context.Context, input DeployRancherInput) {
 	if input.RancherImageTag != "" {
 		values["rancherImageTag"] = input.RancherImageTag
 	}
+	if input.RancherIngressClassName != "" {
+		values["ingress.ingressClassName"] = input.RancherIngressClassName
```
Comment (Danil-Grigorev) [inline test/testenv/rancher.go:199]: Shouldn't rancher pod use the only ingress class available in the cluster by default? We install ngrok or nginx but never both in different scenarios, and I think this `RancherIngressClassName` is not needed.

File: test/testenv/setupcluster.go [hunk 5]
```diff
@@ -69,15 +70,15 @@ func SetupTestCluster(ctx context.Context, input SetupTestClusterInput) *SetupTe
 	Expect(input.ClusterctlConfigPath).ToNot(BeEmpty(), "ClusterctlConfigPath is required for setupTestCluster")
 	Expect(input.Scheme).ToNot(BeNil(), "Scheme is required for setupTestCluster")
 	Expect(input.ArtifactFolder).ToNot(BeEmpty(), "ArtifactFolder is required for setupTestCluster")
-	Expect(input.Hostname).ToNot(BeEmpty(), "Hostname is required for setupTestCluster")
+	//Expect(input.Hostname).ToNot(BeEmpty(), "Hostname is required for setupTestCluster")
```
Comment (Danil-Grigorev) [inline test/testenv/setupcluster.go:73]:
```suggestion
```

File: test/testenv/gitea.go [hunk 6]
```diff
@@ -117,26 +118,57 @@ func DeployGitea(ctx context.Context, input DeployGiteaInput) *DeployGiteaResult
...
+			return err
+		}
+		defer resp.Body.Close()
+
+		if resp.StatusCode != http.StatusOK {
+			return fmt.Errorf("expected status OK, got %v", resp.Status)
+		}
+
+		return nil
+	}, input.ServiceWaitInterval...).Should(Succeed())
```
Comment (Danil-Grigorev) [inline test/testenv/gitea.go:169]:
```suggestion
  Eventually(func(g Gomega) {
    resp, err := http.Get(url)
    defer resp.Body.Close()
    g.Expect(err).ToNot(HaveOccured())
    g.Expect(resp.StatusCode).To(Equal(http.StatusOK))
  }, input.ServiceWaitInterval...).Should(Succeed())
```

File: test/testenv/service.go [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+		if len(svc.Status.LoadBalancer.Ingress) == 0 {
+			return errors.New("No load balancer ingress")
+		}
+
+		if svc.Status.LoadBalancer.Ingress[0].Hostname == "" {
+			return errors.New("No service host name")
+		}
+
+		return nil
+	}, input.IngressWaitInterval...).Should(Succeed(), "Failed to get service host name")
```
Comment (Danil-Grigorev) [inline test/testenv/service.go:79]:
```suggestion
  Eventually(func(g Gomega) error {
    g.Expect(komega.Get(svc)()).To(Succeed())
    g.Expect(svc.Status.LoadBalancer.Ingress).ToNot(HaveLen(0))
    g.Expect(svc.Status.LoadBalancer.Ingress[0].Hostname).ToNot(Equal(""))
  }, input.IngressWaitInterval...).Should(Succeed(), "Failed to get service host name")
```

File: test/e2e/suites/v2prov/suite_test.go [hunk 8]
```diff
@@ -94,10 +104,10 @@ var _ = BeforeSuite(func() {
 		ClusterctlConfigPath: clusterctlConfigPath,
 		Scheme:               e2e.InitScheme(),
 		ArtifactFolder:       flagVals.ArtifactFolder,
-		Hostname:             hostName,
-		KubernetesVersion:    e2eConfig.GetVariable(e2e.KubernetesManagementLegacyVersionVar),
+		KubernetesVersion:    e2eConfig.GetVariable(e2e.KubernetesManagementVersionVar),
```
[the remark this answers is not in the document]
Comment (Danil-Grigorev) [reply test/e2e/suites/v2prov/suite_test.go:107]: We don’t need to restart rancher to fix v1alpha3 CRDs there, as the test case was using 1.4.6 CRDs for CAPIProvider too.

[the remark this answers is not in the document]
Comment (richardcase) [reply test/e2e/suites/v2prov/suite_test.go:107]: We're only going to support the CAPI extension from v2.8.x

File: test/e2e/suites/embedded-capi-disabled/suite_test.go [hunk 9]
```diff
@@ -142,27 +175,79 @@ var _ = BeforeSuite(func() {
...
+		Image:                        fmt.Sprintf("ghcr.io/rancher/turtles-e2e-%s", runtime.GOARCH),
 		Tag:                          "v0.0.1",
 		WaitDeploymentsReadyInterval: e2eConfig.GetIntervals(setupClusterResult.BootstrapClusterProxy.GetName(), "wait-controllers"),
 		AdditionalValues: map[string]string{
 			"cluster-api-operator.cert-manager.enabled":      "false",
 			"rancherTurtles.features.embedded-capi.disabled": "false",
 		},
+	}
+	if flagVals.UseEKS {
+		rtInput.AdditionalValues["rancherTurtles.imagePullSecrets"] = "{regcred}"
```
[the remark this answers is not in the document]
Comment (richardcase) [reply test/e2e/suites/embedded-capi-disabled/suite_test.go:205]: No mistake, thats how you need to pass it as its a list.

File: charts/rancher-turtles/Chart.lock [hunk 10]
```diff
@@ -1,6 +1,6 @@
 dependencies:
 - name: cluster-api-operator
   repository: https://kubernetes-sigs.github.io/cluster-api-operator
-  version: 0.5.1
```
[the remark this answers is not in the document]
Comment (richardcase) [reply charts/rancher-turtles/Chart.lock:4]: I have removed the lock file and added a gitignore.

File: .github/workflows/run-e2e-suite.yaml [hunk 11]
```diff
@@ -30,62 +37,73 @@ env:
   AZURE_CLIENT_ID: ${{ secrets.AZURE_CLIENT_ID }}
   AZURE_CLIENT_SECRET: ${{ secrets.AZURE_CLIENT_SECRET }}
   AZURE_TENANT_ID: ${{ secrets.AZURE_TENANT_ID }}
+  GITHUB_USERNAME: $${{ github.actor }}
```
[the remark this answers is not in the document]
Comment (richardcase) [reply .github/workflows/run-e2e-suite.yaml]: :moneybag: 
Removed

File: test/e2e/suites/update-labels/suite_test.go [hunk 12]
```diff
@@ -138,26 +131,86 @@ var _ = BeforeSuite(func() {
...
+		BootstrapClusterProxy:        setupClusterResult.BootstrapClusterProxy,
+		HelmBinaryPath:               flagVals.HelmBinaryPath,
+		ChartPath:                    flagVals.ChartPath,
+		CAPIProvidersYAML:            e2e.CapiProviders,
+		Namespace:                    turtlesframework.DefaultRancherTurtlesNamespace,
+		Image:                        fmt.Sprintf("ghcr.io/rancher/turtles-e2e-%s", runtime.GOARCH),
+		Tag:                          "v0.0.1",
+		WaitDeploymentsReadyInterval: e2eConfig.GetIntervals(setupClusterResult.BootstrapClusterProxy.GetName(), "wait-controllers"),
+		AdditionalValues: map[string]string{
+			"cluster-api-operator.cluster-api.version":          "v1.6.0",
```
[the remark this answers is not in the document]
Comment (richardcase) [reply test/e2e/suites/update-labels/suite_test.go:196]: Turtles by default will support v1.4.x. This suite is explicitly testing a newer version of CAPI to make sure we don't break v2prov.

This test will only be needed until the change we added to Rancher adds the required labels.

File: test/testenv/setupcluster.go (hunk 5 above)
Comment (richardcase) [reply test/testenv/setupcluster.go]: Removed.

**Review by Danil-Grigorev — APPROVED** [review]
I think at this point it is better to get it merged and debug in local env later. I’ll approve it. My concern was that we are trying to make too many changes in this PR. If the e2e tests continue to fail, it might be issue with EKS configuration or a re-occurring problem with wrong CAPI CRDs version. But again, we will figure it out.

**Review by furkatgofurov7 — APPROVED** [review]

**Comment by richardcase** [comment]
We may have to do some tweaks to secrets and workflows.

I also agree as a follow up we need to refactor to do:
- Split into full, short and local suites (reduce the if/then conditions)
- Make the `BeforeSuite` composable so we don't repeat the same code.

**Pull request merged**